tachyon/math/base/groups.h

```diff
@@ -74,6 +84,110 @@ class MultiplicativeGroup : public MultiplicativeSemigroup<G> {
     G ret = *static_cast<const G*>(this);
     return ret.InverseInPlace();
   }
+
+  template <typename Container>
+  constexpr static bool BatchInverseInPlace(Container& groups,
+                                            const G& coeff = G::One()) {
+    return BatchInverse(groups, &groups, coeff);
+  }
+
+  // This is taken and modified from
+  // https://github.com/arkworks-rs/algebra/blob/5dfeedf560da6937a5de0a2163b7958bd32cd551/ff/src/fields/mod.rs#L355-L418.
+  // Batch inverse: [b₁, b₂, ..., bₙ] = [a₁⁻¹, a₂⁻¹, ... , aₙ⁻¹]
+  template <typename InputContainer, typename OutputContainer>
+  constexpr static bool BatchInverse(const InputContainer& groups,
+                                     OutputContainer* inverses,
+                                     const G& coeff = G::One()) {
+    if (std::size(groups) != std::size(*inverses)) {
+      LOG(ERROR) << "Size of |groups| and |inverses| do not match";
+      return false;
+    }
+
+#if defined(TACHYON_HAS_OPENMP)
+    using G2 = decltype(std::declval<G>().Inverse());
+    size_t thread_nums = static_cast<size_t>(omp_get_max_threads());
+    if (std::size(groups) >=
+        size_t{1} << (thread_nums / kParallelBatchInverseDivisorThreshold)) {
+      size_t num_elem_per_thread =
+          (std::size(groups) + thread_nums - 1) / thread_nums;
+
+      auto groups_chunks = base::Chunked(groups, num_elem_per_thread);
+      auto inverses_chunks = base::Chunked(*inverses, num_elem_per_thread);
+      auto zipped = base::Zipped(groups_chunks, inverses_chunks);
+      auto zipped_vector = base::Map(
+          zipped.begin(), zipped.end(),
+          [](const std::tuple<absl::Span<const G2>, absl::Span<G2>>& v) {
+            return v;
+          });
+
+#pragma omp parallel for
+      for (size_t i = 0; i < zipped_vector.size(); ++i) {
+        const auto& [fields_chunk, inverses_chunk] = zipped_vector[i];
+        DoBatchInverse(fields_chunk, inverses_chunk, coeff);
+      }
+      return true;
+    }
+#endif
+    DoBatchInverse(absl::MakeConstSpan(groups), absl::MakeSpan(*inverses),
+                   coeff);
+    return true;
+  }
+
+ private:
+  // NOTE(chokobole): This value was chosen empirically that
+  // |batch_inverse_benchmark| performs better at fewer input compared to the
+  // number of cpu cores.
+  constexpr static size_t kParallelBatchInverseDivisorThreshold = 4;
+
+  FRIEND_TEST(GroupsTest, BatchInverse);
+  constexpr static void DoBatchInverse(const absl::Span<const G>& groups,
+                                       absl::Span<G> inverses, const G& coeff) {
+    // Montgomery’s Trick and Fast Implementation of Masked AES
+    // Genelle, Prouff and Quisquater
+    // Section 3.2
+    // but with an optimization to multiply every element in the returned
+    // vector by |coeff|.
+
+    // First pass: compute [a₁, a₁ * a₂, ..., a₁ * a₂ * ... * aₙ]
```

Comment:
It may start from a0.

Reply:
@chokobole Is it?

Reply:
When seeing the comment `BatchInverse` above, it looks to me no problem! Could you check it again @dongchangYoo ?

Reply:
It seems I misunderstood.!